### DISCUSSION TITLE
Local and system headers files project selector

### DISCUSSION THREAD

**Discussion by rossanoparis** [KeyWorksRW/wxUiEditor#1434] [Code Generation]
It would be very helpful to add a way based on which the user should be able to add files without using "files selector".

The current file selector, adds files making them relatives to wxUI project path, but in case I have to move a project from a folder to another all #includes directives referred to external libraries become messy, and it is an hard job make them tidy again.

Generally speacking, #include directives for files which belong to a C++ projet can be expressed using their relative paths reffered to wxUI project and that's fine, but for files belonging to external libraries, they are often, in my case always, expressed using relative paths referred to directories defined in IDE tools such as VS, CodeLite, ecc ...

Actually the user is able to get it only typing its own files directly into "local_hdr_includes" or  "system_hdr_includes" fields.
But as you can imagine, when such files are many, it is not confortable.

As I see it, another button to add a string inserted by the user would be helpful also a resizable dialog would be useful.
![image](https://github.com/KeyWorksRW/wxUiEditor/assets/33904565/2288c0e6-049d-4a00-8c4b-33863e79bb2c)

**Comment by Randalphwa** [comment]
Tonight's 4/23 build will change what happens when you click the Add button in the dialog that first comes up when you click the button in either the system_src or system_hdr properties. I'd appreciate it if you could try this out when you get a chance and let me know if this will ultimately make it easier for you to add header files from other libraries.

**Comment by rossanoparis** [comment]
Thank you @Randalphwa I just tried your last modification. 
You did more than I would have expected, it would have been enough a way to insert free text by teh user ...

Anyway what you did, works perfectly.

Perhaps, if I were you, I'd avoid to use wxWidgwets include folder as default, because it takes really to long to show the content.
I'd suggest to show a "busy cursor" while the SW is getting files from a given folder.

Is there a way to remove a folder from the list?
![image](https://github.com/KeyWorksRW/wxUiEditor/assets/33904565/ea455114-4fe8-49b5-a1b1-b7d5b2a5394c)

**Comment by Randalphwa** [comment]
Both busy cursor and removal button is planned. On Windows the entries are stored under:

```HKEY_CURRENT_USER\Software\KeyWorks\wxUiEditor\preferences\sys_header_dlg```

Short-term, you can use regedit to remove until I get the Remove functionality added.

I'm leaning towards a Search edit box with auto-complete for really large libraries like wxWidgets/include. In some cases the dev will already know the name of the file to insert, but might not know or simply not want to scroll to the section that has the folder containing header file. Even auto-scrolling to a specific folder would be helpful.

_If_ I leave in the wxWidgets directory (only available if $WXWIN is set) then I'll add some smarts to it to remove directories you probably won't care about. E.g., if you are running on Windows, then the osx/, unix/, and /univ directories proably aren't going to be useful.

BTW, the wxWidgets directory is only added if you have < 9 folders in the history for that drop down. The same is true for the root directory your project is in -- that's only added if you have < 9 folders in the history for that drop down.

**Comment by rossanoparis** [comment]
Kind @Randalphwa, in my opinion, what you are doing requires an effort that probably won't cover all expectations.

For instance, I normally use libraries based on .hpp files or in other cases I neeed to include other kind of files such as .cc, which are usefull to split up an extra large class ...

Not all a developer needs to "include", is a .h header.
That's why, if I were you, I'd let the developer inserts a free text without the "automation" you are going to implement.
A part all you are implementing, could you let the user inserts a free text too?

Don't get me wrong, but I'm afraid you are doing more than it is necessary.

Just to clarify what I think about.
My own libraries folder, contains 245 headers files grouped in 45 folders.
Certainly, to select the one I need, it is necessary a filter to find out the right one.
This imply I have to know what I need, so why should I type a filter instead of type directly the file a need.
I think it is easier for a developer types directly something like "wxhControls/wxhButton.h" which will be inserted as `#include <wxhControls/wxhButton.h>`

... perhaps it could be a mix of what you are implementing and the free content of "header_preamble" field.

**Comment by Randalphwa** [comment]
It probably makes sense to put a toolbar above that list that handles add, edit, remove, move up and move down, and then the side buttons would be Find... and Sort. That would make it look and function more like `wxArrayStringProperty`. That would be the same for both system and local header files.
